### PR TITLE
[IMP] loyalty,_*: loyalty in portal

### PR DESCRIPTION
**Pull request by dere-odoo**
_* = sale_loyalty, pos_loyalty
Prior to this commit:
- When a customer has a loyalty card or ewallet
- He is not able to find any information about it apart from checkout (or
  PoS ticket)
- He never has access to the history of transactions.
- He does not have a digital card that he can easily find and use while
  purchasing in PoS (through mobile)

Post this commit:
- Added modals on the portal so that customers can access loyalty cards and
  e-wallets, showing the summary for each card.
- On the portal loyalty card modal, added a barcode for POS. Added a "top up
   now" option for e-wallets.
- Added a loyalty history model to track changes in loyalty cards.
- Added a new controller to show the history of a given coupon.
- Added a loyalty card summary table on the sale order form and the sale portal
  form.

Task-3793399

See also:
odoo/enterprise#67003

### CODE REVIEW

**Review by kcv-odoo — COMMENTED** [review]
Initial review :see_no_evil:

**Review by Feyensv — CHANGES_REQUESTED** [review]
partial review, but I think I'll let @kcv-odoo finish his review first as it doesn't seem to be ready yet (didn't notice directly that he already posted a review).

File: addons/loyalty/static/src/js/loyalty_card_modal.js [hunk 1]
```diff
@@ -0,0 +1,43 @@
...
+publicWidget.registry.PortalLoyaltyWidget = publicWidget.Widget.extend({
+    selector: '.o_loyalty_card_modal_selector',
+    events: {
+        'click .o_loyalty_button_click': '_onPortalLoyalty',
+    },
+    showHistory: true,
+    showRewards: false,
+    async _loadData(ev){
+        let card_id = ev.currentTarget.dataset.card
+        let program = ev.currentTarget.dataset.program
```
Comment (kcv-odoo) [inline addons/loyalty/static/src/js/loyalty_card_modal.js:15]:
```suggestion
        let card_id = ev.currentTarget.dataset.card;
        let program = ev.currentTarget.dataset.program;
```
why let not const :thinking:

File: addons/loyalty/static/src/js/loyalty_card_modal.js [hunk 2]
```diff
@@ -0,0 +1,43 @@
...
+        let result = await rpc("/my/loyalty_portal_values", {"card_id": card_id});
+        result = {
+            ...result,
+            'showRewards': this.showRewards,
+            'showHistory': this.showHistory,
+            'program_type': program,
+            'img_path': program == "Loyalty Cards"
```
Comment (kcv-odoo) [inline addons/loyalty/static/src/js/loyalty_card_modal.js:22]: You should not depend on translatable string to check condition :see_no_evil:

File: addons/loyalty/static/src/js/loyalty_card_modal.js [hunk 3]
```diff
@@ -0,0 +1,43 @@
...
+                ? '/loyalty/static/src/img/Award.svg'
+                : '/loyalty/static/src/img/Wallet.svg',
```
Comment (kcv-odoo) [inline addons/loyalty/static/src/js/loyalty_card_modal.js:24]: File name should in small letter

File: addons/loyalty/views/loyalty_card_views.xml [hunk 4]
```diff
@@ -11,13 +11,30 @@
...
+                                name="%(loyalty_update_balance_wizard_action)d"
+                                class="p-0 text-info fw-normal"
+                                type="action"
+                            >
+                                <field name="points_display" nolabel="1"/>
+                            </button>
                         </group>
                     </group>
+                    <notebook>
+                        <page string="History">
```
Comment (kcv-odoo) [inline addons/loyalty/views/loyalty_card_views.xml:25]:
```suggestion
                        <page string="history_page">
```

File: addons/loyalty/views/loyalty_card_views.xml [hunk 5]
```diff
@@ -11,13 +11,30 @@
...
+                            <field name="history_ids" >
```
Comment (kcv-odoo) [inline addons/loyalty/views/loyalty_card_views.xml:26]:
```suggestion
                            <field name="history_ids">
```

File: addons/loyalty/views/loyalty_card_views.xml [hunk 6]
```diff
@@ -11,13 +11,30 @@
...
+                                <tree edit="0" delete="0" create="0">
+                                    <field name="description"></field>
+                                    <field name="order_id"></field>
+                                    <field name="date"></field>
+                                    <field name="issued"></field>
+                                    <field name="used"></field>
+                                    <field name="new_balance"></field>
```
Comment (kcv-odoo) [inline addons/loyalty/views/loyalty_card_views.xml:33]:
```suggestion
                                    <field name="description"/>
                                    <field name="order_id"/>
                                    <field name="date"/>
                                    <field name="issued"/>
                                    <field name="used"/>
                                    <field name="new_balance"/>
```

File: addons/loyalty/views/loyalty_portal_template.xml [hunk 7]
```diff
@@ -0,0 +1,94 @@
+<odoo>
+    <template id="loyalty_buttons" inherit_id="portal.side_content">
+        <xpath expr="//div[@id='portal_contact']" position="after">
```
Comment (kcv-odoo) [inline addons/loyalty/views/loyalty_portal_template.xml:3]:
```suggestion
        <div id="portal_contact" position="after">
```

File: addons/loyalty/views/loyalty_portal_template.xml [hunk 8]
```diff
@@ -0,0 +1,94 @@
...
+                  <tr class="active">
+                    <th>Description</th>
+                    <th>Order</th>
+                    <th>Date</th>
+                    <th>Issued</th>
+                    <th>Used</th>
+                    <th>New Balance</th>
+                  </tr>
+              </thead>
+                <tbody>
```
Comment (kcv-odoo) [inline addons/loyalty/views/loyalty_portal_template.xml:71]:
```suggestion
              </thead>
              <tbody>
```
Also you should check indent throughout the file

**Review by kcv-odoo — CHANGES_REQUESTED** [review]
Posting a second review so you can start working on the changes :see_no_evil:

File: addons/website_sale_loyalty/models/sale_order.py [hunk 9]
```diff
@@ -3,7 +3,7 @@
 from collections import defaultdict
 from datetime import timedelta
 
-from odoo import api, fields, models
+from odoo import api, fields, models, _
```
Comment (kcv-odoo) [inline addons/website_sale_loyalty/models/sale_order.py:6]:
```suggestion
from odoo import _, api, fields, models
```

File: addons/sale_loyalty/views/sale_portal_templates.xml [hunk 10]
```diff
@@ -62,4 +62,57 @@
...
+            <attribute name="class" add="d-flex gap-2" separator=" "/>
+        </div>
+        <xpath expr="//div[@id='total']/*" position="before">
+            <div class="col-xs-5 col-md-6 col-lg-5">
+               <t t-call='sale_loyalty.sale_order_loyalty_table_template' />
+            </div>
+        </xpath>
+    </template>
+
+    <record id="sale_order_form_loyalty_values" model="ir.ui.view">
```
Comment (kcv-odoo) [inline addons/sale_loyalty/views/sale_portal_templates.xml:105]: sale order view should not be in templates files

File: addons/sale_loyalty/views/sale_portal_templates.xml [hunk 11]
```diff
@@ -62,4 +62,57 @@
...
+        <field name="name">sale.order.form.loyalty.view</field>
+        <field name="model">sale.order</field>
+        <field name="inherit_id" ref="sale.view_order_form"></field>
```
Comment (kcv-odoo) [inline addons/sale_loyalty/views/sale_portal_templates.xml:108]:
```suggestion
        <field name="inherit_id" ref="sale.view_order_form"/>
```

File: addons/sale_loyalty/views/sale_portal_templates.xml [hunk 12]
```diff
@@ -62,4 +62,57 @@
...
+        </table>
+    </template>
+
+    <template id="sale_order_portal_loyalty_values" inherit_id="sale.sale_order_portal_content">
+        <div id="total" position="attributes">
+            <attribute name="class" add="d-flex gap-2" separator=" "/>
+        </div>
+        <xpath expr="//div[@id='total']/*" position="before">
+            <div class="col-xs-5 col-md-6 col-lg-5">
+               <t t-call='sale_loyalty.sale_order_loyalty_table_template' />
```
Comment (kcv-odoo) [inline addons/sale_loyalty/views/sale_portal_templates.xml:100]:
```suggestion
               <t t-call='sale_loyalty.sale_order_loyalty_table_template'/>
```

File: addons/sale_loyalty/views/sale_portal_templates.xml [hunk 13]
```diff
@@ -62,4 +62,57 @@
         </xpath>
     </template>
 
+    <template id="sale_order_loyalty_table_template">
+        <table class="table table-sm" t-if="sale_order.state == 'sale' and loyalty_card">
+            <tr class="border-0">
+                <td><strong>Loyalty Card</strong></td>
+                <td class="text-end">
+                    <span t-out="loyalty_card['old_balance']"/> <span class="d-none d-md-inline" t-out="point_name"/>
```
Comment (kcv-odoo) [inline addons/sale_loyalty/views/sale_portal_templates.xml:70]: should be in two lines
(same for other lines)

File: addons/sale_loyalty/static/src/components/loyalty_card/loyalty_card.js [hunk 14]
```diff
@@ -0,0 +1,32 @@
...
+        this.props.loyalty_card = loyalty_card_values["loyalty_card"]  || undefined;
+        this.props.point_name = loyalty_card_values['point_name']
+    }
+}
+
+export const loyaltyCardComponent = {
+    component: LoyaltyCardComponent,
+};
+
+registry.category("fields").add("loyalty-card-values", loyaltyCardComponent);
```
Comment (kcv-odoo) [inline addons/sale_loyalty/static/src/components/loyalty_card/loyalty_card.js:32]: Why this is in field category :thinking: 
You should create view_widget as this is not related to specific field IMO

File: addons/website_sale_loyalty/views/loyalty_portal_template.xml [hunk 15]
```diff
@@ -0,0 +1,11 @@
+<odoo>
+    <template id="webiste_sale_loyalty_button" inherit_id="loyalty.loyalty_buttons">
+        <p id="card_points" position="after">
+            <t t-if="program=='Loyalty Cards'" >
+                <t t-if="any(reward.required_points &lt;= card.points for reward in card.program_id.reward_ids)">
+                    <p class="m-0 text-info">A reward is waiting for you</p>
```
[the remark this answers is not in the document]
Comment (dere-odoo) [reply addons/website_sale_loyalty/views/loyalty_portal_template.xml:6]: Yes, Because the use now button in the card is shown only when website_sale is installed (which redirects to /shop)
so when website_sale is installed user can check the card and click on claim now

File: addons/loyalty/static/src/js/loyalty_card_modal_template.xml [hunk 16]
```diff
@@ -0,0 +1,110 @@
...
+                    <t t-if="props.history_lines.length">
+                        <strong>Last Transactions</strong>
+                        <div t-if="line_index&lt;5"
+                            t-foreach="props.history_lines"
+                            t-as="line"
+                            t-key="line_index"
+                            class="row px-2">
+                            <div class="col-7">
+                                <t t-if="line.order_id">
+                                    <a t-if="line.order_name == 'sale.order'" t-att-href="line.order_portal_url">
```
Comment (kcv-odoo) [inline addons/loyalty/static/src/js/loyalty_card_modal_template.xml:35]: sale.order here :thinking:

File: addons/loyalty/views/portal_template.xml [hunk 17]
```diff
@@ -0,0 +1,97 @@
...
+                      <th>Description</th>
+                      <th>Order</th>
+                      <th>Date</th>
+                      <th>Issued</th>
+                      <th>Used</th>
+                      <th>New Balance</th>
+                    </tr>
+                </thead>
+                <tbody>
+                    <t t-foreach="coupon_history_lines" t-as="move">
```
Comment (kcv-odoo) [inline addons/loyalty/views/portal_template.xml:75]: why move?

File: addons/loyalty/views/portal_template.xml [hunk 18]
```diff
@@ -0,0 +1,97 @@
...
+                      <tr>
+                            <td><span t-out="move.description"></span></td>
+                            <td t-if="move.order_id">
+                                <a t-if="move.order_id._name == 'sale.order'" t-att-href="move.order_id.sudo().get_portal_url()">
```
Comment (kcv-odoo) [inline addons/loyalty/views/portal_template.xml:79]: may be this giving missing record error :thinking: 
image
image
when history does not have order field set

File: addons/loyalty/views/portal_template.xml [hunk 19]
```diff
@@ -0,0 +1,97 @@
...
+                            </div>
+                        </div>
+                    </div>
+                </div>
+            </div>
+        </div>
+    </template>
+
+    <template id="portal_my_home_loyalty_history" name="Portal layout : loyalty history moves"
+        inherit_id="portal.portal_breadcrumbs">
```
Comment (kcv-odoo) [inline addons/loyalty/views/portal_template.xml:44]:
```suggestion
    <template
        id="portal_my_home_loyalty_history"
        name="Portal layout : loyalty history moves"
        inherit_id="portal.portal_breadcrumbs"
    >
```

File: addons/loyalty/wizard/loyalty_generate_wizard.py [hunk 20]
```diff
@@ -28,6 +28,7 @@ class LoyaltyGenerateWizard(models.TransientModel):
     valid_until = fields.Date()
     will_send_mail = fields.Boolean(compute='_compute_will_send_mail')
     confirmation_message = fields.Char(compute='_compute_confirmation_message')
+    description = fields.Text(string="Descripton", default='Gift for customer')
```
Comment (kcv-odoo) [inline addons/loyalty/wizard/loyalty_generate_wizard.py:31]: No need to add default values may be placeholder is better and add default value if empty

This way user don't have to every time remove this values and add his preferred values

File: addons/loyalty/wizard/loyalty_generate_wizard.py [hunk 21]
```diff
@@ -83,5 +84,13 @@ def generate_coupons(self):
             customers = wizard._get_partners() or range(wizard.coupon_qty)
             for partner in customers:
                 coupon_create_vals.append(wizard._get_coupon_values(partner))
-        self.env['loyalty.card'].create(coupon_create_vals)
+        card = self.env['loyalty.card'].create(coupon_create_vals)
```
Comment (kcv-odoo) [inline addons/loyalty/wizard/loyalty_generate_wizard.py:87]:
```suggestion
        coupons = self.env['loyalty.card'].create(coupon_create_vals)
```

File: addons/loyalty/wizard/loyalty_generate_wizard_views.xml [hunk 22]
```diff
@@ -32,6 +32,10 @@
                             </span>
                             <field name="valid_until"/>
                         </group>
+                        <group class="d-flex flex-column" >
```
Comment (kcv-odoo) [inline addons/loyalty/wizard/loyalty_generate_wizard_views.xml:35]:
```suggestion
                        <group class="d-flex flex-column">
```

File: addons/loyalty/wizard/loyalty_history_wizard.py [hunk 23]
```diff
@@ -0,0 +1,35 @@
+from odoo import _, fields, models
+from odoo.exceptions import ValidationError
+
+
+class SaleLoyaltyHistoryWizard(models.TransientModel):
+    _name = 'sale.loyalty.history.wizard'
+    _description = 'History Coupons'
```
Comment (kcv-odoo) [inline addons/loyalty/wizard/loyalty_history_wizard.py:7]:
```suggestion
    _description = "History Coupons"
```

File: addons/loyalty/wizard/loyalty_history_wizard.py [hunk 24]
```diff
@@ -0,0 +1,35 @@
...
+
+    card_id = fields.Many2one(
+        comodel_name='loyalty.card',
+        required=True,
+        default=lambda self: self.env.context.get('active_id')
```
Comment (kcv-odoo) [inline addons/loyalty/wizard/loyalty_history_wizard.py:12]:
```suggestion
        default=lambda self: self.env.context.get('active_id'),
```
same for other fields :see_no_evil:

File: addons/loyalty/wizard/loyalty_history_wizard.py [hunk 25]
```diff
@@ -0,0 +1,35 @@
+from odoo import _, fields, models
+from odoo.exceptions import ValidationError
+
+
+class SaleLoyaltyHistoryWizard(models.TransientModel):
+    _name = 'sale.loyalty.history.wizard'
```
Comment (kcv-odoo) [inline addons/loyalty/wizard/loyalty_history_wizard.py:6]: why `sale` word in loyalty module :thought_balloon:

File: addons/loyalty/wizard/loyalty_history_wizard_view.xml [hunk 26]
```diff
@@ -0,0 +1,38 @@
...
+                    <group>
+                        <field name="old_balance"/>
+                        <field name="new_balance"/>
+                        <field name="description"/>
+                    </group>
+                </sheet>
+                <footer>
+                    <button special="cancel" data-hotkey="x">
+                        Cancel
+                    </button>
```
Comment (kcv-odoo) [inline addons/loyalty/wizard/loyalty_history_wizard_view.xml:19]: Confirm button should be before cancel button

File: addons/loyalty/models/loyalty_card.py [hunk 27]
```diff
@@ -41,6 +41,7 @@ def _compute_display_name(self):
 
     use_count = fields.Integer(compute='_compute_use_count')
     active = fields.Boolean(default=True)
+    history_ids = fields.One2many(comodel_name='loyalty.history', inverse_name='card_id')
```
Comment (kcv-odoo) [inline addons/loyalty/models/loyalty_card.py:44]: this field should be readonly :thinking: 
because currently it is allowing to update history_line
image

File: addons/loyalty/views/loyalty_card_views.xml [hunk 28]
```diff
@@ -11,13 +11,30 @@
...
                             <field name="expiration_date"/>
                             <field name="partner_id"/>
-                            <label string="Balance" for="points"/>
-                            <span class="d-inline-block">
-                                <field name="points" class="w-auto oe_inline me-1"/>
-                                <field name="point_name" no_label="1" class="d-inline"/>
-                            </span>
+                            <label string="Balance" for="points_display"/>
+                            <button
+                                name="%(loyalty_update_balance_wizard_action)d"
```
Comment (kcv-odoo) [inline addons/loyalty/views/loyalty_card_views.xml:16]: You should create object method in python to pass default values like old balance
or context to in action method to set default values in update balance wizard

File: addons/loyalty/security/ir.model.access.csv [hunk 29]
```diff
@@ -5,3 +5,8 @@ access_loyalty_program,access_loyalty_program,model_loyalty_program,base.group_u
 access_loyalty_reward,access_loyalty_reward,model_loyalty_reward,base.group_user,0,0,0,0
 access_loyalty_rule,access_loyalty_rule,model_loyalty_rule,base.group_user,0,0,0,0
 access_loyalty_generate_wizard,access_loyalty_generate_wizard,model_loyalty_generate_wizard,base.group_user,0,0,0,0
+access_loyalty_history,access_loyalty_history,model_loyalty_history,base.group_user,1,1,1,0
+portal_access_loyalty_history,portal_access_loyalty_history,model_loyalty_history,base.group_portal,1,0,0,0
+access_sale_loyalty_history_wizard, access_sale_loyalty_history_wizard,model_sale_loyalty_history_wizard,base.group_user,1,1,1,0
+access_loyalty_history_admin,access_loyalty_history_admin,model_loyalty_history,base.group_erp_manager,1,1,1,1
+access_sale_loyalty_history_wizard_admin,access_sale_loyalty_history_wizard_admin,model_sale_loyalty_history_wizard,base.group_erp_manager,1,1,1,1
```
Comment (kcv-odoo) [inline addons/loyalty/security/ir.model.access.csv:12]: empty line missing

**Review by kcv-odoo — COMMENTED** [review]
Thanks for the huge work and changes! :sparkles: 
Added some comments.
There are still some things to check. I will try to help you tomorrow so hopefully it will be ready to review before EOW :crossed_fingers:

**Review by kcv-odoo — COMMENTED** [review]
Almost done!

Main issues:
- Top-up-wallet will only work if auto confirm is enabled or have to confirm related SO manually to update wallet amount, Need to add this functionality for POS (after freeze maybe)
- The way history lines created and updated might cause issues need to improve way of history lines computation
Other then that LGTM

@Feyensv Can you take look at other things while we try to improve history lines computation a bit?

File: addons/pos_loyalty/models/loyalty_history.py [hunk 30]
```diff
@@ -0,0 +1,9 @@
+# Part of Odoo. See LICENSE file for full copyright and licensing details.
+
+from odoo import models, fields
+
+
+class LoyaltyHistory(models.Model):
+    _inherit = "loyalty.history"
```
Comment (Feyensv) [inline addons/pos_loyalty/models/loyalty_history.py:7]:
```suggestion
    _inherit = 'loyalty.history'
```

File: addons/sale_loyalty/static/src/views/fields/loyalty_total/loyalty_total_field.js [hunk 31]
```diff
@@ -0,0 +1,27 @@
+import { registry } from '@web/core/registry';
```
Comment (Feyensv) [inline addons/sale_loyalty/static/src/views/fields/loyalty_total/loyalty_total_field.js:1]:
```suggestion
import { registry } from "@web/core/registry";
```

File: addons/sale_loyalty/tests/test_loyalty_history.py [hunk 32]
```diff
@@ -0,0 +1,90 @@
+from odoo.fields import Command
+from odoo.tests.common import tagged
+
+from odoo.addons.sale_loyalty.tests.common import TestSaleCouponCommon
+
+
+@tagged('post_install', '-at_install')
+class TestLoyatyhistory(TestSaleCouponCommon):
```
Comment (Feyensv) [inline addons/sale_loyalty/tests/test_loyalty_history.py:8]:
```suggestion
class TestLoyaltyhistory(TestSaleCouponCommon):
```

File: addons/sale_loyalty/tests/test_loyalty_history.py [hunk 33]
```diff
@@ -0,0 +1,90 @@
...
+
+    def setUp(self):
+        super().setUp()
```
Comment (Feyensv) [inline addons/sale_loyalty/tests/test_loyalty_history.py:11]: Please override setUpClass instead of setUp to avoid triggering the setup once by test method.

File: addons/sale_loyalty/views/sale_order_views.xml [hunk 34]
```diff
@@ -30,6 +30,11 @@
             <xpath expr="//tree//field[@name='tax_id']" position="attributes">
                 <attribute name="readonly" add="is_reward_line" separator=" or "/>
             </xpath>
+            <group name='note_group' position="after">
```
Comment (Feyensv) [inline addons/sale_loyalty/views/sale_order_views.xml:33]:
```suggestion
            <group name="note_group" position="after">
```

File: addons/sale_loyalty/views/sale_portal_templates.xml [hunk 35]
```diff
@@ -60,6 +95,13 @@
                 <t t-set="hide_intro" t-value="1"/>
             </t>
         </xpath>
+        <div id="total" position="attributes">
+            <attribute name="class" add="d-flex gap-2" separator=" "/>
+        </div>
+        <div id="total" position="before">
+            <t t-call="sale_loyalty.sale_order_portal_loyalty_content">
+            </t>
```
Comment (Feyensv) [inline addons/sale_loyalty/views/sale_portal_templates.xml:103]:
```suggestion
            <t t-call="sale_loyalty.sale_order_portal_loyalty_content"/>
```

**Review by Feyensv — CHANGES_REQUESTED** [review]
Second tech review, didn't test for now.

File: addons/loyalty/views/portal_template.xml [hunk 36]
```diff
@@ -0,0 +1,96 @@
...
+        <div name="portal_contact" position="after">
+            <div class='o_loyalty_container'>
+                <div
+                    t-foreach="cards_per_programs"
+                    t-as="program"
+                    class="portal-loyalty-buttons cursor-pointer"
+                >
+                    <div
+                        t-att-data-card="card.id"
+                        t-att-data-program="program"
```
[the remark this answers is not in the document]
Comment (dere-odoo) [reply addons/loyalty/views/portal_template.xml]: Yes, when clicked on this div it sends the data to js component [here](https://github.com/odoo/odoo/pull/173476/files#diff-005a8aee8e8e52e6c55729e0222db933204871f142ed53d40ecc25c98069a0d0R13
although the program parameter is not required so i have removed it

File: addons/loyalty/static/src/js/portal/loyalty_card_dialog/loyalty_card_dialog.js [hunk 37]
```diff
@@ -0,0 +1,12 @@
+import { Component } from '@odoo/owl';
+import { Dialog } from '@web/core/dialog/dialog';
+
+export class PortalLoyaltyCardDialog extends Component {
+    static components = { Dialog };
+    static template = 'loyalty.portal_loyalty_card_dialog';
+    static props = ['*'];
```
[the remark this answers is not in the document]
Comment (dere-odoo) [reply addons/loyalty/static/src/js/portal/loyalty_card_dialog/loyalty_card_dialog.js:7]: Here we dont need any props for the component to work. But when originally i didn't add the static props line, I had pylint error so i added this line. As for widget props, I think widget is for backend_assets, I tried adding it but it gave me bundle error as this component is frontend component

File: addons/loyalty/controllers/portal.py [hunk 38]
```diff
@@ -0,0 +1,117 @@
...
+                "program_type": program_type,
+            },
+            'history_lines': [{
+                'order_id': line.order_id,
+                'description': line.description,
+                'order_portal_url': line.get_order_portal_url(),
+                'points': f'''
+                    {'-' if line.issued < line.used else '+'}
+                     {card_sudo._format_points(abs(line.issued - line.used))}
+                ''',
```
[the remark this answers is not in the document]
Comment (dere-odoo) [reply addons/loyalty/controllers/portal.py:106]: We did this because of the line was > 100 cols, and this seems to work

File: addons/sale_loyalty/models/sale_order.py [hunk 39]
```diff
@@ -43,6 +44,70 @@ def _compute_reward_total(self):
...
+            aggregates=['new_balance:sum', 'issued:sum', 'used:sum'],
+        )
+        order_loyalty_mapping = {
+            data[0]: {
+                'total_old_balance': data[1] - data[2] + data[3],
+                'total_issued': data[2],
+                'total_cost': data[3],
+                'total_new_balance': data[1],
+            }
+            for data in loyalty_data
```
Comment (kcv-odoo) [inline addons/sale_loyalty/models/sale_order.py:91]:
```suggestion
            for order_ref, new_bal, issued, cost in loyalty_data
```
To have better readability :thinking:

File: addons/sale_loyalty/models/sale_order.py [hunk 40]
```diff
@@ -43,6 +44,70 @@ def _compute_reward_total(self):
...
+
+    @api.depends('order_line', 'state')
+    def _compute_loyalty_data(self):
+        confirmed_so = self.filtered(lambda order: order.state == 'sale' and order.id)
+        (self - confirmed_so).loyalty_data = {}
+        if not confirmed_so:
+            return
+
+        order_list = ['sale.order,%i' % (order.id) for order in confirmed_so]
+        loyalty_data = self.env['loyalty.history'].sudo()._read_group(
```
Comment (kcv-odoo) [inline addons/sale_loyalty/models/sale_order.py:77]:
```suggestion
        loyalty_history_data = self.env['loyalty.history'].sudo()._read_group(
```

File: addons/sale_loyalty/models/sale_order.py [hunk 41]
```diff
@@ -43,6 +44,70 @@ def _compute_reward_total(self):
...
+            domain=[
+                ('order_id', 'in', order_list)
+            ],
+            groupby=['order_id'],
+            aggregates=['new_balance:sum', 'issued:sum', 'used:sum'],
+        )
+        order_loyalty_mapping = {
```
Comment (kcv-odoo) [inline addons/sale_loyalty/models/sale_order.py:84]:
```suggestion
        loyalty_history_data_per_order = {
```

File: addons/sale_loyalty/models/sale_order.py [hunk 42]
```diff
@@ -43,6 +44,70 @@ def _compute_reward_total(self):
...
+            }
+            for data in loyalty_data
+        }
+        for order in confirmed_so:
+            order_reference = 'sale.order,%i' % (order.id)
+            if order_reference not in order_loyalty_mapping:
+                order.loyalty_data = {}
+                continue
+            coupons = order.coupon_point_ids.coupon_id
+            coupon_point_name = (coupons and coupons[0].point_name) or 'Points'
```
Comment (kcv-odoo) [inline addons/sale_loyalty/models/sale_order.py:99]:
```suggestion
            coupon_point_name = (len(coupons) == 1 and coupons.point_name) or _("Points")
```
Display coupon point name if there is only one coupon else points also translation

**Review by kcv-odoo — COMMENTED** [review]
Small nitpicking comment to recent changes :see_no_evil: 
(I didn't check test cases though)

File: addons/pos_loyalty/models/pos_order.py [hunk 43]
```diff
@@ -52,6 +52,27 @@ def validate_coupon_programs(self, point_changes, new_codes):
...
+    def add_loyalty_history_lines(self, coupon_data, coupon_updates):
+        id_mapping = {item['old_id']: int(item['id']) for item in coupon_updates}
+        history_lines_to_create = []
+        for coupon in coupon_data:
+            if int(coupon['card_id']) not in id_mapping:
+                continue
+            card_id = id_mapping[int(coupon['card_id'])]
+            issued = coupon['won']
+            cost = coupon['spent']
+            if (issued or cost) and card_id > 0:
```
[the remark this answers is not in the document]
Comment (dere-odoo) [reply addons/pos_loyalty/models/pos_order.py:64]: In pos_loyalty for this case:
there is no existing loyalty card for a customer
there is a sale quotation for this customer
go to a pos session and settle the order
this will create a card from both sides sale_loyalty and pos_loyalty
so for this case i believe that they delete the first generated card and for this card the id_mapping will have -ve card_id

File: addons/sale_loyalty/controllers/portal.py [hunk 44]
```diff
@@ -0,0 +1,40 @@
...
+    @http.route(['/topup'], type='http', auth="public", website=True, sitemap=False)
+    def topup_wallet(self, product_id):
+        order = request.env['sale.order'].sudo().create({
+            'partner_id': request.env.user.partner_id.id,
+            'order_line': [
+                Command.create({
+                    'product_id': int(product_id),
+                })
+            ]
+        })
```
Comment (Feyensv) [inline addons/sale_loyalty/controllers/portal.py:27]: A public route creating a new order in sudo, with any product id given ?
This seems like a nice way to bloat a database :eyes:

image

**Comment by Feyensv** [comment]
@robodoo r+

**Comment by Feyensv** [comment]
@robodoo r+

squashed the two commits
